### PR TITLE
Reapply "Dynamic view sizing" (#140165)

### PR DESCRIPTION
**Pull request by goderbauer**
This reverts commit https://github.com/flutter/flutter/commit/d24c01bd0c41331bd17165e0173b24c5d05d7c0a.

The original change was reverted because it caused some apps to get stuck on the splash screen on some phones.

An investigation determined that this was due to a rounding error. Example: The device reports a physical size of 1008.0 x 2198.0 with a dpr of 1.912500023841858. Flutter would translate that to a logical size of 527.0588169589221 x 1149.2810314243163 and use that as the input for its layout algorithm. Since the constraints here are tight, the layout algorithm would determine that the resulting logical size of the root render object must be 527.0588169589221 x 1149.2810314243163. Translating this back to physical pixels by applying the dpr resulted in a physical size of 1007.9999999999999 x 2198.0 for the frame. Android now rejected that frame because it didn't match the expected size of 1008.0 x 2198.0 and since no frame had been rendered would never take down the splash screen.

Prior to dynamically sized views, this wasn't an issue because we would hard-code the frame size to whatever the requested size was.

Changes in this PR over the original PR:

* The issue has been fixed now by constraining the calculated physical size to the input physical constraints which makes sure that we always end up with a size that is acceptable to the operating system.
* The `ViewConfiguration` was refactored to use the slightly more convenient `BoxConstraints` over the `ViewConstraints` to represent constraints. Both essentially represent the same thing, but `BoxConstraints` are more powerful and we avoid a couple of translations between the two by translating the` ViewConstraints` from the `FlutterView` to `BoxConstraints` directly when the `ViewConfiguration` is created.

All changes over the original PR are contained in the second commit of this PR.

Fixes b/316813075
Part of https://github.com/flutter/flutter/issues/134501.

### CODE REVIEW

**Comment by goderbauer** [comment]
FYI @jiahaog @chingjun @ditman

File: packages/flutter/test/widgets/view_test.dart
Comment (gspencergoog) [inline packages/flutter/test/widgets/view_test.dart:453]: Why no leak tracking?

File: packages/flutter/lib/src/rendering/view.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/rendering/view.dart:6]: If the `ViewConfiguration` doesn't use `ViewConstraints` anymore, do we need the distinction still? It doesn't appear to be used anywhere in the framework (although it is used in the engine).

Comment (goderbauer) [reply packages/flutter/lib/src/rendering/view.dart:6]: We still need it for the type used by `FlutterView.physicalConstraints` in the engine. This is similar to how the engine uses ViewPadding that the framework then translates in to EdgeInsets at its earliest convenience because the second one is more expressive.

File: packages/flutter/test/widgets/view_test.dart
Comment (goderbauer) [reply packages/flutter/test/widgets/view_test.dart:453]: Between reverting and relanding this change the testWidgetsWithLeakTracking method was removed in https://github.com/flutter/flutter/pull/140239. It's functionality is now folded into `testWidgets` directly.

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Comment by goderbauer** [comment]
Waiting for confirmation on b/316813075 before merging this.

**Pull request merged**
